src/driver/register.rs

```diff
@@ -0,0 +1,130 @@
+use crate::driver;
+
+use libc::{iovec, UIO_MAXIOV};
+use std::cell::RefCell;
+use std::cmp;
+use std::io;
+use std::mem;
+use std::ptr;
+use std::rc::Rc;
+use std::slice;
+
+pub(crate) struct Buffers {
+    raw_vecs: ptr::NonNull<iovec>,
+    buf_states: Vec<BufState>,
+    orig_cap: usize,
+}
+
+enum BufState {
+    Free { init_len: usize },
+    CheckedOut,
+}
+
+impl Buffers {
+    pub(crate) fn new(bufs: impl Iterator<Item = Vec<u8>>) -> Self {
+        let bufs = bufs.take(cmp::min(UIO_MAXIOV as usize, 65_536));
+        let (size_hint, _) = bufs.size_hint();
+        let mut iovecs = Vec::with_capacity(size_hint);
+        let mut buf_states = Vec::with_capacity(size_hint);
+        for mut buf in bufs {
+            iovecs.push(iovec {
+                iov_base: buf.as_mut_ptr() as *mut _,
+                iov_len: buf.capacity(),
+            });
+            buf_states.push(BufState::Free {
+                init_len: buf.len(),
+            });
+            mem::forget(buf);
+        }
+        debug_assert_eq!(iovecs.len(), buf_states.len());
+        let raw_vecs = unsafe { ptr::NonNull::new_unchecked(iovecs.as_mut_ptr()) };
```

Comment:
Is the assumption that iovecs is contiguous not erroneous? I don't believe this is enforced by `impl Iterator`

Reply:
^ Please disregard this